valhalla/midgard/gridded_data.h

```diff
@@ -100,14 +116,325 @@ template <class coord_t> class GriddedData : public Tiles<coord_t> {
    *
    * @return contour line geometries with the larger intervals first (for rendering purposes)
    */
-  contours_t GenerateContours(const std::vector<float>& contour_intervals,
+  contours_t GenerateContours(std::vector<contour_interval_t>& intervals,
                               const bool rings_only = false,
                               const float denoise = 1.f,
-                              const float generalize = 200.f) const;
+                              const float generalize = 200.f) const {
+    // sort the contours first on the metric index then on the values with the bigger contours first
+    std::sort(intervals.begin(), intervals.end(), std::greater<>());
+
+    // Values at tile corners and center (0 element is center)
+    int sh[5];
+    typename PointLL::first_type s[5]; // Values at the tile corners and center
+    PointLL tile_corners[5];           // PointLL at tile corners and center
+
+    // Find the intersection along a tile edge
+    auto intersect = [&tile_corners, &s](int p1, int p2) {
+      auto ds = s[p2] - s[p1];
+      return PointLL((s[p2] * tile_corners[p1].x() - s[p1] * tile_corners[p2].x()) / ds,
+                     (s[p2] * tile_corners[p1].y() - s[p1] * tile_corners[p2].y()) / ds);
+    };
+
+    // which metrics do we need contours for
+    auto _ = std::make_pair(intervals.cbegin(), intervals.cend());
+    std::vector<decltype(_)> metrics{std::move(_)};
+    for (auto interval = intervals.cbegin(); interval != intervals.cend(); ++interval) {
+      if (std::get<0>(*interval) != std::get<0>(*metrics.back().first)) {
+        metrics.back().second = interval;
+        metrics.emplace_back(interval, intervals.cend());
+      }
+    }
+
+    // we need something to hold each iso-line
+    contours_t contours(intervals.size(), std::list<feature_t>{feature_t{}});
+
+    // and something to find them quickly
+    using contour_lookup_t = std::unordered_map<PointLL, typename feature_t::iterator>;
+    std::vector<contour_lookup_t> lookups(intervals.size());
+    // TODO: preallocate the lookups for each interval
+
+    int tile_inc[4] = {0, 1, this->ncolumns_ + 1, this->ncolumns_};
+    int case_value;
+    int case_table[3][3][3] = {{{0, 0, 8}, {0, 2, 5}, {7, 6, 9}},
+                               {{0, 3, 4}, {1, 3, 1}, {4, 3, 0}},
+                               {{9, 6, 7}, {5, 2, 0}, {8, 0, 0}}};
+    // For each metric we tracked
+    for (const auto& metric : metrics) {
```

Comment:
this loop wasnt here before since there was only one metric, time. now we have this external loop to loop over all the metrics for which this grid was created.